kylin/src/main/java/org/apache/zeppelin/kylin/KylinInterpreter.java

```diff
@@ -166,28 +165,42 @@ public String getSQL(String cmd) {
   }
 
   private InterpreterResult executeQuery(String sql) throws IOException {
-
     HttpResponse response = prepareRequest(sql);
+    String result;
 
-    if (response.getStatusLine().getStatusCode() != 200) {
-      logger.error("failed to execute query: " + response.getEntity().getContent().toString());
-      return new InterpreterResult(InterpreterResult.Code.ERROR,
-          "Failed : HTTP error code " + response.getStatusLine().getStatusCode());
-    }
-
-    BufferedReader br = new BufferedReader(
-        new InputStreamReader((response.getEntity().getContent())));
-    StringBuilder sb = new StringBuilder();
+    try {
+      int code = response.getStatusLine().getStatusCode();
+      result = IOUtils.toString(response.getEntity().getContent(), "UTF-8");
+
+      if (code != 200) {
+        StringBuilder errorMessage = new StringBuilder("Failed : HTTP error code " + code);
+        logger.error("Failed to execute query: " + result);
+
+        KylinErrorResponse kylinErrorResponse = KylinErrorResponse.fromJson(result);
+        if (kylinErrorResponse == null) {
+          logger.error("Cannot get json from string: " + result);
+          // when code is 401, the response is html, not json
+          if (code == 401) {
+            errorMessage.append(". Error message: Unauthorized. This request requires "
+                + "HTTP authentication. Please make sure your have set your credentials"
+                + " correctly.");
+          } else {
+            errorMessage.append(". Error message: " + result);
+          }
+        } else {
+          String exception = kylinErrorResponse.getException();
+          logger.error("The exception is " + exception);
+          errorMessage.append(". Error message: " + exception);
+        }
 
-    String output;
-    logger.info("Output from Server .... \n");
-    while ((output = br.readLine()) != null) {
-      logger.info(output);
-      sb.append(output).append('\n');
+        return new InterpreterResult(InterpreterResult.Code.ERROR, errorMessage.toString());
+      }
+    } catch (NullPointerException | IOException e) {
```

Comment:
NullPointerException shouldn't happen then, if kylinErrorResponse == null is checked?

Reply:
@felixcheung yes, NPE shouldn't happen now.